### PR TITLE
Assignment1 combfilter peer review

### PR DESCRIPTION
**Pull request by JohnKasun**

### CODE REVIEW

**Comment by johnnymac647** [comment]
This project looks great! I love the way you've organized your comb filter classes and it's very clear how they all relate to one another. The test cases all seem to be working out well and the plots look great! Well done!

I only have 2 minor questions/suggestions. Firstly, it looks like all of the data is still being written to a text file rather than a new audio file. I know I struggled with this part of the assignment (and low-key still don't have it completely worked out), but if you have the time to do so, I think it would be good to be able to hear the results of the comb filter directly by creating an output wav file in the main function.

Secondly, it doesn't look like your test cases are interacting with the code in your main function that much, and are treated as kind of their own things. I think they get the job doe as far as testing whether the filters themselves behave as expected, but there's a lot of repeated lines regarding memory allocation in each test. Is there a way to break the main function down into some smaller functions (maybe like a create/find input signal function alongside a process/apply filter function), and then have those called in both the main function and the test cases? Also, and this is bordering on nit-picky so I apologize, but could you add some arguments to your test functions for things like sample rate or signal frequency? That way it would be easier to run those same tests on a variety of different signals.

**Comment by JohnKasun** [comment]
> This project looks great! I love the way you've organized your comb filter classes and it's very clear how they all relate to one another. The test cases all seem to be working out well and the plots look great! Well done!
> 
> I only have 2 minor questions/suggestions. Firstly, it looks like all of the data is still being written to a text file rather than a new audio file. I know I struggled with this part of the assignment (and low-key still don't have it completely worked out), but if you have the time to do so, I think it would be good to be able to hear the results of the comb filter directly by creating an output wav file in the main function.
> 
> Secondly, it doesn't look like your test cases are interacting with the code in your main function that much, and are treated as kind of their own things. I think they get the job doe as far as testing whether the filters themselves behave as expected, but there's a lot of repeated lines regarding memory allocation in each test. Is there a way to break the main function down into some smaller functions (maybe like a create/find input signal function alongside a process/apply filter function), and then have those called in both the main function and the test cases? Also, and this is bordering on nit-picky so I apologize, but could you add some arguments to your test functions for things like sample rate or signal frequency? That way it would be easier to run those same tests on a variety of different signals.

Thanks, John! I actually totally missed the "output to wave file" part of the assignment, so I appreciate you catching that!

File: assignment1.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+![](images/MajorTomDiff.PNG)
```
Comment (med1844) [inline assignment1.md:1]: Maybe better to add description or explanation of why there're irregular peaks in difference for `MajorTom16.wav`

File: src/CombFilter/CombFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+//=================================
+Error_t CCombFilterFIR::process(float** ppfAudioInputBuffer, float** ppfAudioOutputBuffer, int iNumberOfFrames)
+{
+	for (int channel = 0; channel < m_iNumChannels; channel++)
+	{
+		CRingBuffer<float>* fCurrentDelayLine = m_fDelayLine[channel];
+		for (int sample = 0; sample < iNumberOfFrames; sample++)
+		{
+			float fCurrentSample = ppfAudioInputBuffer[channel][sample];
+			fCurrentDelayLine->putPostInc(fCurrentSample);
```
Comment (med1844) [inline src/CombFilter/CombFilter.cpp]: You are putting elements here first. When the delay line is full, it will overwrite the value at `iReadPosition`, as shown in function [`putPostInc`](https://github.com/JohnKasun/2022-MUSI6106/blob/assignment1_combfilter_PeerReview/src/CombFilter/RingBuffer.h#L36). You might want to add position overlap detection, or simply change the order of putting and getting elements.

File: src/CombFilter/CombFilter.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+#include "RingBuffer.h"
+
+
+class CCombFilterBase : public CCombFilterIf
+{
+public:
+	CCombFilterBase(float fMaxDelayLengthInS, float fSampleRateInHz, int iNumChannels);
+	virtual ~CCombFilterBase();
+
+	virtual Error_t process(float** ppfAudioInputBuffer, float** ppfAudioOutputBuffer, int iNumberOfFrames) = 0;
```
Comment (med1844) [inline src/CombFilter/CombFilter.h:16]: I especially like the pure virtual function definition here, which refuses others from initializing the base class that actually does not contain any signal processing functionality.

File: src/MUSI6106Exec/MUSI6106Exec.cpp [hunk 4]
```diff
@@ -87,15 +123,24 @@ int main(int argc, char* argv[])
...
     // get audio data and write it to the output text file (one column per channel)
+    // apply comb filter
     while (!phAudioFile->isEof())
     {
         // set block length variable
         long long iNumFrames = kBlockSize;
 
         // read data (iNumOfFrames might be updated!)
         phAudioFile->readData(ppfAudioData, iNumFrames);
+        phCombFilter->process(ppfAudioData, ppfAudioData, iNumFrames);
```
Comment (med1844) [inline src/MUSI6106Exec/MUSI6106Exec.cpp:146]: The same issue as John has mentioned that you forgot to add something like `phOutputAudioFile->writeData(ppfOutputAudioData, iNumFrames)`. This is also part of the assignment requirement.

File: src/MUSI6106Exec/MUSI6106Exec.cpp [hunk 5]
```diff
@@ -127,13 +173,451 @@ int main(int argc, char* argv[])
...
+{
+    bool test1passed = test1();
+    test2();
+    bool test3passed = test3();
+    bool test4passed = test4();
+    bool test5passed = test5();
+}
+
+//Display Input and Output Audio Buffers
+void displayIOBuffers(float** ppfAudioInputBuffer, float** ppfAudioOutputBuffer, int iNumChannels, int iNumSamples)
```
Comment (med1844) [inline src/MUSI6106Exec/MUSI6106Exec.cpp]: This function is not dry. You are actually copy and pasting exactly the same code in [`displayBuffer`](https://github.com/JohnKasun/2022-MUSI6106/blob/assignment1_combfilter_PeerReview/src/MUSI6106Exec/MUSI6106Exec.cpp#L229), with only the content of `std::cout << "Buffer: " << std::endl;` changed. This creates redundant code and should be avoided. You can add a wrapper function like:

```c++
void displayBufferAndName(float** ppfAudioBuffer, int iNumChannels, int iNumSamples, const std::string& bufferName) {
    std::cout << bufferName << ": " << std::endl;
    displayBuffer(ppfAudioBuffer, iNumChannels, iNumSamples);
}
```

And rewrite `displayIOBuffers` as:

```c++
    void displayIOBuffers(float** ppfAudioInputBuffer, float** ppfAudioOutputBuffer, int iNumChannels, int iNumSamples) {
        displayBufferAndName(ppfAudioInputBuffer, iNumChannels, iNumSamples, "Input");
        displayBufferAndName(ppfAudioOutputBuffer, iNumChannels, iNumSamples, "Output");
    }
```

File: src/MUSI6106Exec/MUSI6106Exec.cpp [hunk 6]
```diff
@@ -127,13 +173,451 @@ int main(int argc, char* argv[])
...
+    }
+    delete[] ppfInputBuffer;
+    delete[] ppfOutputBuffer;
+    delete[] ppfMagnitudeRatios;
+
+}
+
+//Helper function for processing a block of audio data with comb filter
+//iStartSample is the offset corresponding to the current read/write position of the original buffer
+void processBlock(float** ppfInputBuffer, float** ppfOutputBuffer, CCombFilterIf* phCombFilter, int iNumChannels, int iStartSample, int iBlockSize)
```
Comment (med1844) [inline src/MUSI6106Exec/MUSI6106Exec.cpp:365]: You can use tricks to simplify the implementation of this function and improve efficiency.

Array is essentially the pointer to the head of the memory region representing the whole array. If you add an offset to the pointer, e.g. `ppfInputBuffer[channel] + 7`, it is equivalent of `&ppfInputBuffer[channel][7]`. This address can also be viewed as the head of memory region of an array, here it is `ppfInputBuffer[channel][7:]` when represented as slicing in python. This can be assigned to `ppfTempInBuffer` and `ppfTempOutBuffer` to eliminate redundant copy and pasting `for` loops as they are essentially processing on the same memory region as the output.

You can rewrite your code as:

```c++
for (int i = 0; i < iNumChannels; ++i) {
  ppfTempInBuffer[i] = ppfInputBuffer[i] + iStartSample;
  ppfTempOutBuffer[i] = ppfOutputBuffer[i] + iStartSample;
  phCombFilter->process(ppfTempInBuffer, ppfTempOutBuffer, iBlockSize);
}
```

You might also want to check [`memoryview`](https://docs.python.org/3/library/stdtypes.html#memoryview) for deeper understanding.

File: src/MUSI6106Exec/MUSI6106Exec.cpp [hunk 7]
```diff
@@ -127,13 +173,451 @@ int main(int argc, char* argv[])
...
+bool test4();
+bool test5();
+
+void runTests()
+{
+    bool test1passed = test1();
+    test2();
+    bool test3passed = test3();
+    bool test4passed = test4();
+    bool test5passed = test5();
```
Comment (med1844) [inline src/MUSI6106Exec/MUSI6106Exec.cpp]: Seems that there's no way to tell if tests are passed or not from the `stdout` or `stderr`? I'm not sure if this is intentional design but it might help if it can print whether each test is passed or not.

**Review by med1844 — COMMENTED** [review]
In general I would say this is a pretty outstanding implementation, with detailed and concise comments and edge-case detection & handling. The naming convention follows Alex's original codebase's style and is easy to read & understand. Some bugs and possible improvements are mentioned above with code snippets.

I also doubt if we really need to let test cases interact with main functions. Unit tests are more about testing if the provided interfaces are working correctly. Parameters like sample rate, signal frequency, signal input should be configured in extra configure files and should not be set in the main function, where codes are expected to **solely** focus on implementing functions of processing audio files.

Still I would agree that these test cases can be provided with the same initialization utilities. This can help reduce the code a lot, meanwhile not affecting main function.
